File: deck/seed/code/native/browser/runtime/compare.ts
// Deep structural equality over the dynamic value, the same semantics on every backend: identical scalars
// are equal, two lists are equal when their sizes match and every element pair is deep-equal, and anything
// else answers by identity. Kept behavior-identical to the Term implementation it replaced (see
// native/shared/test/compare.tree, 2026-08-31: the body moved into per-backend shims so the rust, swift and
// kotlin builds do not need structural operations over their boxed dynamics).
const compareRuntime = {
  deepEqual(a: unknown, b: unknown): boolean {
    if (a === b) {
      return true
    }

    if (Array.isArray(a) && Array.isArray(b)) {
      if (a.length !== b.length) {
        return false
      }

      for (let i = 0; i < a.length; i++) {
        if (!compareRuntime.deepEqual(a[i], b[i])) {
          return false
        }
      }

      return true
    }

    return false
  },

  contains(list: unknown, value: unknown): boolean {
    if (!Array.isArray(list)) {
      return false
    }

    return list.some(item => compareRuntime.deepEqual(item, value))
  },
  asText(value: unknown): string {
    return typeof value === 'string' ? value : ''
  },
  isTruthy(value: unknown): boolean {
    return Boolean(value)
  },
  numeric(value: unknown): number {
    return typeof value === 'number' ? value : Number.NaN
  },
  above(a: unknown, b: unknown): boolean {
    return compareRuntime.numeric(a) > compareRuntime.numeric(b)
  },
  below(a: unknown, b: unknown): boolean {
    return compareRuntime.numeric(a) < compareRuntime.numeric(b)
  },
  gap(a: unknown, b: unknown): number {
    return Math.abs(compareRuntime.numeric(a) - compareRuntime.numeric(b))
  },
}

File: deck/seed/code/native/kotlin/runtime/compare.kt
// Deep structural equality over Any: scalar equality, two lists element by element, else false. Kotlin's
// == already deep-compares lists of scalars; the recursion covers lists of dynamics.
object compareRuntime {
    fun deepEqual(a: Any?, b: Any?): Boolean {
        if (a == b) return true

        if (a is List<*> && b is List<*>) {
            if (a.size != b.size) return false

            for (i in a.indices) {
                if (!deepEqual(a[i], b[i])) return false
            }

            return true
        }

        return false
    }

    fun numeric(v: Any?): Double = when (v) {
        is Long -> v.toDouble()
        is Int -> v.toDouble()
        is Double -> v
        else -> Double.NaN
    }

    fun contains(list: Any?, value: Any?): Boolean {
        if (list !is List<*>) return false

        return list.any { deepEqual(it, value) }
    }

    fun asText(v: Any?): String = (v as? String) ?: ""

    fun isTruthy(v: Any?): Boolean = when (v) {
        null, Unit -> false
        is Boolean -> v
        is Long -> v != 0L
        is Double -> v != 0.0
        is String -> v.isNotEmpty()
        else -> true
    }

    fun above(a: Any?, b: Any?): Boolean = numeric(a) > numeric(b)

    fun below(a: Any?, b: Any?): Boolean = numeric(a) < numeric(b)

    fun gap(a: Any?, b: Any?): Double = Math.abs(numeric(a) - numeric(b))
}
